streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -587,62 +585,66 @@ private KafkaStreams(final InternalTopologyBuilder internalTopologyBuilder,
 
         // The application ID is a required config and hence should always have value
         processId = UUID.randomUUID();
-        final String clientId = config.getString(StreamsConfig.CLIENT_ID_CONFIG);
+        final String userClientId = config.getString(StreamsConfig.CLIENT_ID_CONFIG);
         final String applicationId = config.getString(StreamsConfig.APPLICATION_ID_CONFIG);
-        if (clientId.length() <= 0) {
+        if (userClientId.length() <= 0) {
             this.clientId = applicationId + "-" + processId;
         } else {
-            this.clientId = clientId;
+            this.clientId = userClientId;
         }
 
         final LogContext logContext = new LogContext(String.format("stream-client [%s] ", clientId));
         this.log = logContext.logger(getClass());
 
-        internalTopologyBuilder.setApplicationId(applicationId);
-        // sanity check to fail-fast in case we cannot build a ProcessorTopology due to an exception
-        internalTopologyBuilder.build(null);
-
-        long cacheSize = config.getLong(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG);
-        if (cacheSize < 0) {
-            cacheSize = 0;
-            log.warn("Negative cache size passed in. Reverting to cache size of 0 bytes.");
-        }
-
-        final StateRestoreListener delegatingStateRestoreListener = new DelegatingStateRestoreListener();
-
-        threads = new StreamThread[config.getInt(StreamsConfig.NUM_STREAM_THREADS_CONFIG)];
         try {
-            stateDirectory = new StateDirectory(
-                config,
-                Time.SYSTEM);
+            stateDirectory = new StateDirectory(config, time);
         } catch (final ProcessorStateException fatal) {
             throw new StreamsException(fatal);
         }
-        streamsMetadataState = new StreamsMetadataState(
-            internalTopologyBuilder,
-            parseHostInfo(config.getString(StreamsConfig.APPLICATION_SERVER_CONFIG)));
 
         final MetricConfig metricConfig = new MetricConfig().samples(config.getInt(StreamsConfig.METRICS_NUM_SAMPLES_CONFIG))
             .recordLevel(Sensor.RecordingLevel.forName(config.getString(StreamsConfig.METRICS_RECORDING_LEVEL_CONFIG)))
             .timeWindow(config.getLong(StreamsConfig.METRICS_SAMPLE_WINDOW_MS_CONFIG), TimeUnit.MILLISECONDS);
         final List<MetricsReporter> reporters = config.getConfiguredInstances(StreamsConfig.METRIC_REPORTER_CLASSES_CONFIG,
-            MetricsReporter.class);
+                MetricsReporter.class);
         reporters.add(new JmxReporter(JMX_PREFIX));
-        metrics = new Metrics(metricConfig, reporters, Time.SYSTEM);
+        metrics = new Metrics(metricConfig, reporters, time);
 
-        GlobalStreamThread.State globalThreadState = null;
+        internalTopologyBuilder.setApplicationId(applicationId);
+
+        // sanity check to fail-fast in case we cannot build a ProcessorTopology due to an exception
+        internalTopologyBuilder.build(null);
```

Comment:
nit: this looks a little awkward IMHO, maybe add a method to `InternalTopologyBuilder`  sth like `ensureCanBuildTopology` although I don't like that name and don't have a better suggestion atm.

Reply:
OK how about I just add a `build()` that implements `build(null)` then? Naming-wise I think it is the same as to "build the whole topology", which should fail if anything happens.